From: Bernd Zeimetz <bernd@bzed.de>
Date: Wed, 3 Nov 2010 21:10:03 +0100
Subject: [PATCH] Add missing -lz.

---
 src/Makefile.am |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -197,7 +197,7 @@ endif
 
 viking_SOURCES = main.c
 
-LDADD           = libviking.a $(PACKAGE_LIBS) @EXPAT_LIBS@ @LIBCURL@ icons/libicons.a
+LDADD           = libviking.a $(PACKAGE_LIBS) @EXPAT_LIBS@ @LIBCURL@ icons/libicons.a -lz
 if WINDOWS
 LDADD += \
 	../win32/installer/pixmaps/viking_icon.o
