Description: IT_PROG_INTLTOOL and AM_GNU_GETTEXT conflict

diff --git a/configure.ac b/configure.ac
index 66fa53b..0e6e762 100644
--- a/configure.ac
+++ b/configure.ac
@@ -22,9 +22,6 @@ AC_DEFINE_UNQUOTED(GETTEXT_PACKAGE,"$GETTEXT_PACKAGE",[Gettext package name])
 
 AM_GLIB_DEFINE_LOCALEDIR(LOCALEDIR)
 
-AM_GNU_GETTEXT_VERSION([0.17])
-AM_GNU_GETTEXT([external])
-
 IT_PROG_INTLTOOL([0.35.0])
 
 AC_SUBST([CONFIG_STATUS_DEPENDENCIES],['$(top_srcdir)/po/LINGUAS'])

